From 34fa9d4408942d7ca9b6148fb1f9fcf4887e1a37 Mon Sep 17 00:00:00 2001
From: Alberto97 <albertop2197@gmail.com>
Date: Mon, 18 May 2015 14:31:45 -0700
Subject: [PATCH 6/9] Add alternate WebView provider package

* Fall back to old config_webViewPackageName value if the alternate
  WebView package is not found.

See also http://review.cyanogenmod.org/#/c/97919/

Change-Id: Ifd33d939e7327c6904a1c461133a4ac6389196e8
---
 core/java/android/webkit/WebViewFactory.java | 17 ++++++++++++++++-
 core/res/res/values/config.xml               |  6 +++++-
 core/res/res/values/symbols.xml              |  1 +
 3 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/core/java/android/webkit/WebViewFactory.java b/core/java/android/webkit/WebViewFactory.java
index 474ef42..9a819a9 100644
--- a/core/java/android/webkit/WebViewFactory.java
+++ b/core/java/android/webkit/WebViewFactory.java
@@ -77,7 +77,14 @@ public final class WebViewFactory {
     private static PackageInfo sPackageInfo;
 
     public static String getWebViewPackageName() {
-        return AppGlobals.getInitialApplication().getString(
+        Application initialApp = AppGlobals.getInitialApplication();
+        String pkg = initialApp.getString(
+                com.android.internal.R.string.config_alternateWebViewPackageName);
+        /* Attempt to use alternate WebView package first */
+        if (isPackageInstalled(initialApp, pkg)) {
+            return pkg;
+        }
+        return initialApp.getString(
                 com.android.internal.R.string.config_webViewPackageName);
     }
 
@@ -413,6 +420,14 @@ public final class WebViewFactory {
         return IWebViewUpdateService.Stub.asInterface(ServiceManager.getService("webviewupdate"));
     }
 
+    private static boolean isPackageInstalled(Context context, String packageName) {
+        try {
+            return context.getPackageManager().getPackageInfo(packageName, 0) != null;
+        } catch (PackageManager.NameNotFoundException e) {
+            return false;
+        }
+    }
+
     private static native boolean nativeReserveAddressSpace(long addressSpaceToReserve);
     private static native boolean nativeCreateRelroFile(String lib32, String lib64,
                                                         String relro32, String relro64);
diff --git a/core/res/res/values/config.xml b/core/res/res/values/config.xml
index 68df6f1..4e9a6e8 100755
--- a/core/res/res/values/config.xml
+++ b/core/res/res/values/config.xml
@@ -1898,9 +1898,13 @@
         string that's stored in 8-bit unpacked format) characters.-->
     <bool translatable="false" name="config_sms_decode_gsm_8bit_data">false</bool>
 
-    <!-- Package name providing WebView implementation. -->
+    <!-- Package name providing default WebView implementation. -->
     <string name="config_webViewPackageName" translatable="false">com.android.webview</string>
 
+    <!-- Package name providing alternate WebView implementation.
+         Fall back to config_webViewPackageName if not available. -->
+    <string name="config_alternateWebViewPackageName" translatable="false">com.google.android.webview</string>
+
     <!-- If EMS is not supported, framework breaks down EMS into single segment SMS
          and adds page info " x/y". This config is used to set which carrier doesn't
          support EMS and whether page info should be added at the beginning or the end.
diff --git a/core/res/res/values/symbols.xml b/core/res/res/values/symbols.xml
index 1055547..b39bb4f 100755
--- a/core/res/res/values/symbols.xml
+++ b/core/res/res/values/symbols.xml
@@ -1961,6 +1961,7 @@
   <java-symbol type="string" name="websearch" />
   <java-symbol type="drawable" name="ic_media_video_poster" />
   <java-symbol type="string" name="config_webViewPackageName" />
+  <java-symbol type="string" name="config_alternateWebViewPackageName" />
 
   <!-- From SubtitleView -->
   <java-symbol type="dimen" name="subtitle_corner_radius" />
-- 
2.7.4

